Add 16A8W linear ops support and test

Ninja91 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #13653
* __->__ #13658
* #13641

- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

Ninja91 committed:
Add 16A8W linear ops support and test

- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

[ghstack-poisoned]

Ninja91 commented:
Updated version of https://github.com/pytorch/executorch/pull/13448. Closing the former.

Ninja91 commented:
@per @digantdesai some of CI is failing but it doesn't appear to be related to code change here.

Ninja91 committed:
Update on "Add 16A8W linear ops support and test"


- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

[ghstack-poisoned]

per reviewed (approved):
Nice! Two small suggestions, otherwise LGTM.

Ninja91 commented on backends/arm/test/ops/test_linear.py:
```diff
@@ -258,3 +262,64 @@ def test_linear_vgf_INT(test_data: torch.Tensor):
+
+
+@common.parametrize("test_data", test_data_rank1_INT | test_data_rank4_INT)
+@pytest.mark.xfail(
+    reason="missing int16 linear ops support; fails at TOSA reference model run with Invalid TOSA graph"
+)
```
Will look into that as a follow up. Based on test you shared: https://git.mlplatform.org/tosa/conformance_tests.git/tree/operators/tensor/fully_connected/fully_connected_6x44_i16xi8_acci48

It looks like the output dtype and bias dtype should be INT48. How do we go about this? 
In our current quantization setup we don't add any quant spec for bias and it gets quantized to INT32. Do you recommend adding BIAS spec to quantize to INT16 as well and try? Or do you want me to change both output and bias dtype to int32/int64 as there's no int48 in Pytorch AFAIK.

per replied:
I have an internal fix for that to handle it in the backend.

Ninja91 commented on backends/arm/operators/op_transpose.py:
```diff
@@ -47,7 +47,7 @@ def define_node(
             self.target,
             [inputs[0], output],
-            [ts.DType.INT8, ts.DType.INT32, ts.DType.FP32],
+            [ts.DType.INT8, ts.DType.INT16, ts.DType.INT32, ts.DType.FP32],
```
Will update.

Ninja91 committed:
Update on "Add 16A8W linear ops support and test"


- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

[ghstack-poisoned]

Ninja91 committed:
Update on "Add 16A8W linear ops support and test"


- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

[ghstack-poisoned]

Ninja91 committed:
Update on "Add 16A8W linear ops support and test"


- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Differential Revision: [D80308822](https://our.internmc.facebook.com/intern/diff/D80308822/)

[ghstack-poisoned]

per commented:
@digantdesai Were was this merged? Can't see it on main... Getting a bit confused by the ghstack stuff 😕

digantdesai commented:
@Ninja91 ?

Ninja91 commented:
> @digantdesai Were was this merged? Can't see it on main... Getting a bit confused by the ghstack stuff 😕

This diff was merged and new PR was created for main branch here: https://github.com/pytorch/executorch/pull/13754/

I see that there are some tests failing. Looks like the change to use support_integer() causes FP tests fail as some ops are still INT.
`
if self.tosa_spec.support_integer():
     valid.dtypes.extend([ts.DType.INT8, ts.DType.INT16])
if self.tosa_spec.support_float():
    valid.dtypes.extend([ts.DType.FP16, ts.DType.FP32])
`

What would be your suggestion?

digantdesai commented:
this didn't merge with main, can you push this to main manually?
